### DISCUSSION TITLE
Grouping styles with selectors

### DISCUSSION THREAD

**Discussion by serkodev** [master-co/css#18] [Ideas]
Currently, for using multiple styles with the same selector, you will have:

```html
<div class="
    font:12_.foo
    font:blue_.foo
    font:yellow_.foo@dark
    bg:red_.foo
    bg:green_.foo@dark
    p:20_.foo
    m:10_.foo
">
```

### Plugin

So I developed a plugin [master-styles-group](https://github.com/serkodev/master-styles-group), after using it you can shorten the code to one line:

```html
<div class="{font:12;_bg:red;_{font:yellow;_bg:green}@dark;_p:20;_m:10}_.foo">
```

In addition, it also compatible with putting selector before group.

```html
<div class="{font:12;_bg:red}@xm"></div>
<div class="@xm{font:12;_bg:red}"></div>
```

In the Master [Discord](https://discord.gg/sZNKpAAAw6), @oberonlai shows his usage with `master-styles-group`.

```html
<ul class="
    d:flex flex-wrap:wrap
    >li{mr:15}
    >li>a{d:block;_color:white; _t:center;_font:12}
    >li>a>span{w:40;_h:40;_bg:#191919}
    ›li>a>span>svg{abs;_top:50%;_left:50%;_translate(-50%,-50%)}
    >li>a.active{color: #FCB900}
    >li>a>span>svg>path{~stroke;.2s;ease}
    >li>a:hover>span>svg>path{stroke:#FCB900}
    >li>a.active>span>svg{stroke:#FCB900}
">
```

### Master Styles native support grouping

After some discussion, @1Aron also agrees that grouping is a useful feature and can reduce the code a lot. Since it may require changes to existing syntax and other breaking changes, it may not be implemented until `styles` v2.0 publish (#17).

I hope you guys can share some views on it.

**Comment by 1aron** [comment]
The **Group** feature has been released in `@master/css@1.16.0`, but the APIs are a bit different.

```html
<div class="p:10|20|30|40:hover@sm text:center:hover@sm font:red-60/.5:hover@sm">without group</div>
<div class="{p:10|20|30|40;text:center;font:red-60/.5}:hover@sm">group</div>
```

Because selectors have different behavior at the start and at the end, it is not possible to put the current selector in front.

**Comment by aileksandar** [comment]
What do you think about some alternative syntax for group instead of curly braces `{}` ?
In JSX and Svelte I constantly need to escape the curly brace and that makes code look a bit ugly 😢  Is there a reason why it can't be regular or angled braces instead?

**Reply by 1aron** [reply]
@meoweloper If you use it with our Master CSS Language, the source code would look like this:

![Screen Shot 2022-08-22 at 8 44 01 PM](https://user-images.githubusercontent.com/33840671/185924004-2a5affb1-4b91-4a86-a31a-402f8b959aaa.jpg)

The syntax for using `{}` as a group is mainly evolved from native CSS, and our philosophy is to try to be the same as native and avoid inventing.

Native CSS
```html
<style>
.card { 
    background-color: white;
    padding: 1rem 1.25rem;
}
</style>
<div class="card">...</div>
```

Master CSS
```html
<div class="bg:white p:16|20">...</div>
or
<div class="{bg:white;p:16|20}">...</div>
```

So you can observe that we transform `{background-color: white; padding: 1rem 1.25rem}` into `{bg:white;p:16|20}`.

**Reply by aileksandar** [reply]
Thanks for the quick reply! 
I totally get it, that is one of the reasons I prefer master over tailwind, it feels like css syntax, just concise. 
How are you approaching the issue with escaping curly brace in your projects? Or are you always using plain html or angular?

**Reply by 1aron** [reply]
@meoweloper Your reason is good enough to be a [Review](https://www.producthunt.com/products/master-css/reviews/new). 😂
Can you give an example code? Since I'm using Angular, React doesn't have a problem with escaping at all.

**Reply by 1aron** [reply]
Using Group Styles in React looks like this:

![Screen Shot 2022-08-22 at 9 31 38 PM](https://user-images.githubusercontent.com/33840671/185933317-61bfd258-81a7-4f02-8654-5f55c5d6647f.jpg)

It works.

**Reply by aileksandar** [reply]
Using it in Svelte requires wrapping it in `{''}`

<img width="552" alt="Screenshot 2022-08-22 at 15 37 47" src="https://user-images.githubusercontent.com/1159476/185934881-0191d315-afc8-4bfc-99d0-fbfb16cf7d54.png">

Not great, not terrible 🤷

**Reply by 1aron** [reply]
@meoweloper OMG. Svelte... I saw and played it. 🥲

**Reply by 1aron** [reply]
All I can say is that Svelte is a little naughty. 😅

**Reply by aileksandar** [reply]
If you like that native feel, you would definitely like Svelte. Until you combine it with group syntax 😄

**Reply by 1aron** [reply]
Alternative 😆
```html
<div class="&#123;text:center;italic&#125;">
```
rendered as:
```html
<div class="{text:center;italic}">
```

**Reply by 1aron** [reply]
I've always followed Svelte and liked it! 💖

**Reply by aileksandar** [reply]
> ```
> <div class="&#123;text:center;italic&#125;">
> ```
This could work if you can make it render non-ugly in the VS Code extension 😄

**Reply by 1aron** [reply]
Great idea, beautifying it seems like a good solution.
